### PR TITLE
Documentation for CREATE SEQUENCE

### PR DESCRIPTION
**Pull request by hectorgcr**

### CODE REVIEW

**Comment by hectorgcr** [comment]
To make it easier to review the code, you can see the new documentation here: http://10.150.0.4:35493/latest/api/ysql/

File: content/latest/api/ysql/_index.md [hunk 1]
```diff
@@ -92,6 +92,14 @@ Statement | Description |
 [`BEGIN`](transactions) | Start a transaction |
 [`END` &#124; `COMMIT`](transactions) | Commit a transaction |
 
+## Sequences
```
Comment (ravimurthy) [inline content/latest/api/ysql/_index.md:95]: How about DROP SEQUENCE?

File: content/latest/api/ysql/create_sequence.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+- `minvalue` is the minimum value allowed in the sequence. If this value is reached (in a sequence with a negative increment), `nextval()` will return an error. If `NO MINVALUE` is specified, the default value will be used. Default is 1.
+- `maxvalue` is the maximum value allowed in the sequence. If this value is reached, `nextval()` will return an error. If `NO MAXVALUE` is specified, the default will be used. Default is 2^63 - 1;
+- `start` is the first value in the sequence. Default is 1.
+- `cache` how many numbers from the sequence to cache in the client. Default is 1.
+
+## Semantics
+- An error is raised if a sequence with that name already exists in the current schema.
+
+## Cache
+In YSQL as in Postgres, the sequence's data is stored in two values: `last_val` and `is_called`. `last_val` stores the last value allocated for that sequence. When `is_called` is false indicates that `last_val` hasn't been used, and the next call to `nextval()` could return that value.
```
Comment (ravimurthy) [inline content/latest/api/ysql/create_sequence.md]: Clarify that sequence data is stored in a persistent system table. The table has one row per sequence - and stores last_val and is_called. List out the two columns and their semantics 
last_val - ...
is_called - ....

File: content/latest/api/ysql/create_sequence.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+By default (when `INCREMENT` is 1), each call to `nextval()` updates `last_val` for that sequence. In YSQL, the table holding the sequence's data is replicated as opposed to being in the local file system. Each update to this table requires two RPCs (a read RPC and a write RPC), so the latency experienced by a single `nextval()` in YSQL will be significantly higher than the same operation in Postgres. To avoid such performance degradation, we recommend using a cache value with a value large enough so that the latency of one allocation can be amortized over all the numbers allocated for the cache.
```
Comment (ravimurthy) [inline content/latest/api/ysql/create_sequence.md]: ... requires two RPCs currently (and will be optimized to 1 RPC in the future). In any case, latency will be significantly higher...

Comment (ravimurthy) [inline content/latest/api/ysql/create_sequence.md]: Mention that the cached values are stored in memory on the localnode, and avoids the RPC.

**Comment by hectorgcr** [comment]
Addressed @ravimurthy comments.
New server running on http://10.150.0.4:42626/latest/api/ysql/

**Comment by hectorgcr** [comment]
Added a note about `ALTER SEQUENCE` and `setval()` not being supported yet

File: content/latest/api/ysql/_index.md [hunk 4]
```diff
@@ -92,6 +92,17 @@ Statement | Description |
...
+Statement | Description |
+----------|------------|
+[`CREATE SEQUENCE`](create_sequence) | Create a new sequence |
+[`DROP SEQUENCE`](drop_sequence) | Drop a sequence |
+[`nextval(sequence)`](nextval_sequence) | Get the next value in the sequence
+[`currval(sequence)`](currval_sequence) | Get the last value returned by a nextval call
```
Comment (ndeodhar) [inline content/latest/api/ysql/_index.md]:
```suggestion
[`currval(sequence)`](currval_sequence) | Get the last value returned by the most recent nextval call for the specified sequence
```

File: content/latest/api/ysql/_index.md [hunk 5]
```diff
@@ -92,6 +92,17 @@ Statement | Description |
...
+[`lastval()`](lastval_sequence) | Get the last value returned by any nextval call
```
Comment (ndeodhar) [inline content/latest/api/ysql/_index.md]:
```suggestion
[`lastval()`](lastval_sequence) | Get the last value returned by the most recent nextval call for any sequence
```

File: content/latest/api/ysql/create_sequence.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+### start
+<svg class="rrdiagram" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="135" height="35" viewbox="0 0 135 35"><path class="connector" d="M0 22h5m125 0h5"/><rect class="literal" x="5" y="5" width="125" height="25" rx="7"/><text class="text" x="15" y="22">&lt;Integer Literal&gt;</text></svg>
+
+### cache
+<svg class="rrdiagram" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="135" height="35" viewbox="0 0 135 35"><path class="connector" d="M0 22h5m125 0h5"/><rect class="literal" x="5" y="5" width="125" height="25" rx="7"/><text class="text" x="15" y="22">&lt;Integer Literal&gt;</text></svg>
+
+
+### Grammar
+```
+create_sequence = 'CREATE' 'SEQUENCE' [ 'IF' 'NOT' 'EXISTS' ] sequence_name sequence_options;
```
Comment (ndeodhar) [inline content/latest/api/ysql/create_sequence.md]: I think we can remove the single quotes around 'CREATE', 'SEQUENCE' etc. - keep it consistent with how we've specified INSERT/SELECT statements.
(Same comment applies below for DROP SEQUENCE also).

File: content/latest/api/ysql/create_sequence.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+sequence_name = '<Text Literal>';
+sequence_options = [ 'INCREMENT' [ 'BY' ] increment ] [ 'MINVALUE' minvalue | 'NO' 'MINVALUE' ] [ 'MAXVALUE' maxvalue | 'NO' 'MAXVALUE' ] [ 'START' [ 'WITH' ] start ] [ 'CACHE' cache ];
```
Comment (ndeodhar) [inline content/latest/api/ysql/create_sequence.md]: START WITH and CACHE are going outside of the gray box when you view this.

File: content/latest/api/ysql/create_sequence.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+```
+
+Where
+
+- `sequence_name` is the name of the sequence.
+- `increment` is the difference between consecutive values in the sequence. Default is 1.
+- `minvalue` is the minimum value allowed in the sequence. If this value is reached (in a sequence with a negative increment), `nextval()` will return an error. If `NO MINVALUE` is specified, the default value will be used. Default is 1.
+- `maxvalue` is the maximum value allowed in the sequence. If this value is reached, `nextval()` will return an error. If `NO MAXVALUE` is specified, the default will be used. Default is 2^63 - 1.
+- `start` is the first value in the sequence. Default is 1.
+- `cache` how many numbers from the sequence to cache in the client. Default is 1.
```
Comment (ndeodhar) [inline content/latest/api/ysql/create_sequence.md]: nit (to keep sentence construction similar):
```suggestion
- `cache` specifies how many numbers from the sequence to cache in the client. Default is 1.
```

File: content/latest/api/ysql/create_sequence.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+## Semantics
+- An error is raised if a sequence with that name already exists in the current schema.
```
Comment (ndeodhar) [inline content/latest/api/ysql/create_sequence.md]:
```suggestion
- An error is raised if a sequence with that name already exists in the current schema and `IF NOT EXISTS` is not specified.
```

File: content/latest/api/ysql/create_sequence.md [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+`SERIAL` types create a sequence with a cache with default value of 1. So `SERIAL` types should be avoided, and their equivalent statement should be used.
+Instead of creating a table with a `SERIAL` type like this:
+```
+CREATE TABLE t(k SERIAL)
+```
+You should create a sequence with a large enough cache first, and then set the column that you want to have a serial type to `DEFAULT` to `nextval()` of the sequence.
+```
+CREATE SEQUENCE t_k_seq CACHE 10000;
+CREATE TABLE t(k integer NOT NULL DEFAULT nextval('t_k_seq');
```
Comment (ndeodhar) [inline content/latest/api/ysql/create_sequence.md]: Missing ending parenthesis.
```suggestion

CREATE TABLE t(k integer NOT NULL DEFAULT nextval('t_k_seq'));
```

File: content/latest/api/ysql/create_sequence.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+SELECT nextval('s2');
+```
+```
+nextval
+---------
+   10001
+(1 row)
+```
+
+Create a sequence that starts at 0. MINVALUE also has to be changed from its default 1 to something less than or equal to 0.
```
Comment (ndeodhar) [inline content/latest/api/ysql/create_sequence.md:148]: This should also be mentioned above where you are describing MINVALUE and START.

File: content/latest/api/ysql/drop_sequence.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+```
+
+Where
+
+- `sequence_name` is the name of the sequence.
+- `CASCADE`. Remove also all objects that depend on this sequence (for example a `DEFAULT` value in a table's column).
+- `RESTRICT`. Do not remove this sequence if any object depends on it. This is the default behavior even if it's not specified.
+
+## Semantics
+- An error is raised if a sequence with that name already exists in the current schema.
```
Comment (ndeodhar) [inline content/latest/api/ysql/drop_sequence.md]:
```suggestion
- An error is raised if a sequence with that name does not exist in the current schema unless `IF EXISTS` is specified.
```

**Review by ndeodhar — CHANGES_REQUESTED** [review]

File: content/latest/api/ysql/currval_sequence.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+### Diagram
+<svg class="rrdiagram" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="315" height="35" viewbox="0 0 315 35"><path class="connector" d="M0 22h5m66 0h10m66 0h10m118 0h10m25 0h5"/><rect class="literal" x="5" y="5" width="66" height="25" rx="7"/><text class="text" x="15" y="22">SELECT</text><rect class="literal" x="81" y="5" width="66" height="25" rx="7"/><text class="text" x="91" y="22">currval(</text><a xlink:href="../grammar_diagrams#sequence-name"><rect class="rule" x="157" y="5" width="118" height="25"/><text class="text" x="167" y="22">sequence_name</text></a><rect class="literal" x="285" y="5" width="25" height="25" rx="7"/><text class="text" x="295" y="22">)</text></svg>
+
+### sequence_name
+<svg class="rrdiagram" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="117" height="35" viewbox="0 0 117 35"><path class="connector" d="M0 22h5m107 0h5"/><rect class="literal" x="5" y="5" width="107" height="25" rx="7"/><text class="text" x="15" y="22">&lt;Text Literal&gt;</text></svg>
+
+### Grammar
+```
+currval ::= SELECT currval'(' sequence_name ')';
```
Comment (ndeodhar) [inline content/latest/api/ysql/currval_sequence.md]: Remove the single quotes similar to what you've done elsewhere.
```suggestion
currval ::= SELECT currval( sequence_name );
```

File: content/latest/api/ysql/lastval_sequence.md [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+The `lastval()` function returns the value returned from the last call to `nextval()` (for any sequence) in the current session.
+
+## Syntax
+
+### Diagram
+<svg class="rrdiagram" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="154" height="35" viewbox="0 0 154 35"><path class="connector" d="M0 22h5m66 0h10m68 0h5"/><rect class="literal" x="5" y="5" width="66" height="25" rx="7"/><text class="text" x="15" y="22">SELECT</text><rect class="literal" x="81" y="5" width="68" height="25" rx="7"/><text class="text" x="91" y="22">lastval()</text></svg>
+
+### Grammar
+```
+lastval ::= SELECT lastval'()';
```
Comment (ndeodhar) [inline content/latest/api/ysql/lastval_sequence.md]:
```suggestion
lastval ::= SELECT lastval();
```

File: content/latest/api/ysql/nextval_sequence.md [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+### Diagram
+<svg class="rrdiagram" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="318" height="35" viewbox="0 0 318 35"><path class="connector" d="M0 22h5m66 0h10m69 0h10m118 0h10m25 0h5"/><rect class="literal" x="5" y="5" width="66" height="25" rx="7"/><text class="text" x="15" y="22">SELECT</text><rect class="literal" x="81" y="5" width="69" height="25" rx="7"/><text class="text" x="91" y="22">nextval(</text><a xlink:href="../grammar_diagrams#sequence-name"><rect class="rule" x="160" y="5" width="118" height="25"/><text class="text" x="170" y="22">sequence_name</text></a><rect class="literal" x="288" y="5" width="25" height="25" rx="7"/><text class="text" x="298" y="22">)</text></svg>
+
+### sequence_name
+<svg class="rrdiagram" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="117" height="35" viewbox="0 0 117 35"><path class="connector" d="M0 22h5m107 0h5"/><rect class="literal" x="5" y="5" width="107" height="25" rx="7"/><text class="text" x="15" y="22">&lt;Text Literal&gt;</text></svg>
+
+### Grammar
+```
+nextval ::= SELECT nextval'(' sequence_name ')';
```
Comment (ndeodhar) [inline content/latest/api/ysql/nextval_sequence.md]:
```suggestion
nextval ::= SELECT nextval( sequence_name );
```

**Review by ndeodhar — CHANGES_REQUESTED** [review]
Minor comments - LGTM otherwise.

**Review by ndeodhar — APPROVED** [review]

**Pull request merged**